universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts

```diff
@@ -1,33 +1,82 @@
 import {BalanceObserver} from './BalanceObserver';
 import {TokenDetailsWithBalance} from '@universal-login/commons';
 import {utils} from 'ethers';
+import {} from '@universal-login/commons/lib';
+import {PriceObserver} from './PriceObserver';
+import clonedeep from 'lodash.clonedeep';
 
-export interface PriceObserver {
-  getTokenPrice: (tokenSymbol: string, currencySymbol: string) => Promise<number>;
-}
+export type SymbolToValue = {[symbol: string]: number};
 
 export class AggregateBalanceObserver {
+  private tokenPrices: {[symbol: string]: SymbolToValue} = {};
+  private tokenDetailsWithBalance: TokenDetailsWithBalance[] = [];
+  private unsubscribePriceObserver?: Function;
+  private unsubscribeBalanceObserver?: Function;
+  private callbacks: Function[] = [];
   constructor(private balanceObserver: BalanceObserver, private priceObserver: PriceObserver) {}
 
-  subscribe(callback: Function, currencySymbol: string) {
-    const callbackWrapper = async (tokensDetailsWithBalance: TokenDetailsWithBalance[]) => {
-      const aggregatedBalance = await this.getAggregatedBalance(tokensDetailsWithBalance, currencySymbol);
-      callback(aggregatedBalance);
+  subscribe(callback: Function) {
+    this.callbacks.push(callback);
+    if (this.callbacks.length === 1) {
+      this.unsubscribeBalanceObserver = this.balanceObserver.subscribe(this.balanceObserverCallback.bind(this));
+      this.unsubscribePriceObserver = this.priceObserver.subscribe(this.priceObserverCallback.bind(this));
+    }
+
+    const unsubscribe = () => {
+      this.callbacks = this.callbacks.filter((element) => callback !== element);
+      if (this.callbacks.length === 0) {
+        this.unsubscribeBalanceObserver!();
+        this.unsubscribePriceObserver!();
+        this.tokenDetailsWithBalance = [];
+        this.tokenPrices = {};
+      }
     };
+    return unsubscribe;
+  }
 
-    return this.balanceObserver.subscribe(callbackWrapper.bind(this));
+  priceObserverCallback(tokenPrices: {}) {
+    this.tokenPrices = tokenPrices;
+    this.notifyListeners();
   }
 
-  async getAggregatedBalance(tokensDetailsWithBalance: TokenDetailsWithBalance[], currencySymbol: string) {
-    return tokensDetailsWithBalance.reduce(
-      async (total, tokenDetails) => (await total + await this.getTokenBalance(tokenDetails, currencySymbol)),
-      Promise.resolve(0)
-    );
+  balanceObserverCallback(tokenDetailsWithBalance: TokenDetailsWithBalance[]) {
+    this.tokenDetailsWithBalance = tokenDetailsWithBalance;
+    this.notifyListeners();
   }
 
-  async getTokenBalance(tokenDetailsWithBalance: TokenDetailsWithBalance, currencySymbol: string) {
+  notifyListeners() {
+    if (Object.keys(this.tokenPrices).length === 0 || this.tokenDetailsWithBalance.length === 0) {
+      return;
+    }
+    const totalWorth = this.getAggregatedTotalWorth();
+    this.callbacks.forEach((callback) => callback(totalWorth));
+  }
+
+  getAggregatedTotalWorth() {
```

Comment:
Let's refactor last 4 functions:
1) You can use mapValues from loadash
2) Introduce new types:
```ts
interface ConverstionMapping {
  symbol: [symbol: BigNumber]
}

class TokenPriceMapping {
  conversions : ConverstionMapping;
  sums(tokenBalances: []);
  total();
}

}

```